File: answers/leetcode/Pascal's Triangle II.cpp
//@result 34 / 34 test cases passed. Status: Accepted Runtime: 0 ms Submitted: 0 minutes ago You are here! Your runtime beats 12.14% of cpp submissions.

class Solution {
public:
    vector<int> getRow(int rowIndex) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		std::vector<int> row;
		row.push_back(1);
		if (0 == rowIndex)
		{
			return row;
		}
		if (1 == rowIndex)
		{
			row.push_back(1);
			return row;
		}
		for (int i = 1; i <= rowIndex; ++ i)
		{
			for (int j = 0; j + 1 < row.size(); ++ j)
			{
				row[j] = row[j] + row[j + 1];
			}
			row.insert(row.begin(), 1);
		}
		return row;
    }
};

File: answers/leetcode/Pascal's Triangle II/Pascal's Triangle II.java
class Solution {
    private List<List<Integer>> triangle;

    public List<Integer> getRow(int rowIndex) {
        this.generate(rowIndex);
        return this.triangle.get(rowIndex);
    }

    private void generate(int numRows) {
        if (null == this.triangle) {
            this.triangle = new ArrayList<List<Integer>>();
            List<Integer> row = new ArrayList<Integer>();
            row.add(1);
            triangle.add(row);
        }
        if (this.triangle.size() > numRows) {
            return;
        }

        for (int i = this.triangle.size(); i <= numRows; i++) {
            List<Integer> row = new ArrayList<Integer>(triangle.get(i - 1));
            row.add(0, 0);
            row.add(0);
            List<Integer> next = new ArrayList<Integer>();
            for (int j = 1; j < row.size(); j++) {
                next.add(row.get(j - 1) + row.get(j));
            }
            triangle.add(next);
        }
    }
}